### DISCUSSION TITLE
User can assign users to own user group even if permission not granted.

### DISCUSSION THREAD

**Discussion by BridgingUnit** [craftcms/cms#8603] [Ideas]
### Description

I've got a SuperUser group who can register users to groups that they have been given permission to add to. However, it also appears they can add users to their own group, even if that's not been explicitly permitted.

### Steps to reproduce

1. Create a user in a User Group "Group A" with the following permissions:
Assign User Groups
Assign users to "Group B"
Assign users to "Group C"

Note: 
Don't check permission to Assign users to "Group A" - so whereas should be able to assign to "Group B" and "Group C", should **not** be able to assign to "Group A".

2. Login in as new user for Group A

3. Create new user. 
Click Permissions tab.
Assignable User Groups includes Group A along with Group B and Group C.
Shouldn't it only show Group B and Group C?

### Additional info

- Craft version: 3.0.22
- May be related to [2820](https://github.com/craftcms/cms/issues/2820), [1188](https://github.com/craftcms/cms/issues/1188), [915](https://github.com/craftcms/cms/issues/915), [2087](https://github.com/craftcms/cms/issues/2087)

**Comment by brandonkelly** [comment]
This is expected behavior (see discussion at #2087).

**Comment by BridgingUnit** [comment]
OK. It's a bit counter intuitive when you're actually assigning permissions though. Any way to make the expected behaviour clearer at that point?

Also, how about reversing the idea? Any way to make it so a user **can't** add a new user to their own user group **unless** it's been explicitly permitted then? Perhaps this is a feature request....

**Comment by BridgingUnit** [comment]
In 2087 you write:

> So to summarize, going forward if you want a user to be allowed to create new users with the same group(s) they belong to, all you will need to do is give them the “Assign user groups” permission (but none of its nested permissions).

But in this instance I **am** making a choice between the nested permissions, or thought I was. The new user isn't being granted carte blanche. I only want them to be able to assign to specific user groups.

**Comment by brandonkelly** [comment]
That comment should be read as, “_at the least_ you just need to give them the ‘Assign user groups’ permission”. I didn’t mean that assigning nested permissions would remove the ability to assign one’s own user groups on other users.

> Also, how about reversing the idea? Any way to make it so a user **can't** add a new user to their own user group **unless** it's been explicitly permitted then? Perhaps this is a feature request....

Craft doesn’t currently have the concept of a “negative permission”. Perhaps we could add a new permission that specifically enables the current behavior – “Assign own user groups” or something – and write a migration that adds that permission to every user that currently has the “Assign user groups” permission, so there’s no change in behavior.

**Comment by BridgingUnit** [comment]
Thanks Brandon. 'Assign own user groups' sounds like a good idea.

I was very surprised to see that by default, when they had permission to assign groups, that a user could assign to their own group, especially given what the interface seemed to be suggesting. Is there any way of making that clearer in the interface do you think? Also I've not checked that this is clear in the docs too.

For an earlier version of Craft 2 I had a plugin which enabled me as superadmin to enable specified user groups to register new users only to specific user groups.

The 'Assign users to Group A/B/C' etc. granular permissions seems to almost cover what I was doing in that regard, but with this new permission 'Assign own user groups' I think it would close the loop. 

Thanks again.

**Comment by kevadamson** [comment]
Any update on this? I don't want users in a group to be able to assign users to their own group. The permissions list seems to suggest this is possible. Seems like an obvious thing to have in place. Let's say a user has made a payment to have that level of access to certain content, yet now can give away that access for free to whomever they want ...

**Comment by kevadamson** [comment]
P.S. I'm also looking for a way of having a user only have access to users in a group that they themselves have added (a bit like the way "other authors" works on entries). I think that and the above would be great enhancements to the "Users" model.

**Comment by Nicovery** [comment]
Hi, 

I have the same issue. 

The permissions interface seems to imply that we can decide which group an user can assign.

In the UsersController there is a condition checking if the user is in the group and the permission to assign to this group. I think we just only need to check the permission. I don't the see the cases where we need to check both.

https://github.com/craftcms/cms/blob/794f21422068e92631b5798076ef62780ea988f1/src/controllers/UsersController.php#L2066-L2072
